## Gather

In [2]:
import pandas as pd
import numpy as np

In [3]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [4]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [4]:
treatments.shape

(280, 7)

In [5]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [6]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [9]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [11]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
patients.sample(5)

patient_id assigned_sex    given_name      surname              address  \
3             4         male          Liêm         Phan  2335 Webster Street   
139         140       female       Novalie         Berg     1275 Goldie Lane   
500         501       female         Jinke    de Keizer    649 Nutter Street   
479         480         male          Tapa  Arsanukayev   4720 Gordon Street   
301         302       female  Onyekachukwu       Obinna   2970 Forest Avenue   

              city       state  zip_code        country  \
3       Woodbridge          NJ    7095.0  United States   
139     Cincinnati          OH   45202.0  United States   
500  Overland Park          MO   64110.0  United States   
479        Ontario  California   91762.0  United States   
301       New York          NY   10004.0  United States   

                                        contact  birthdate  weight  height  \
3    PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
139      NovalieBerg@gustr.com+1 (513) 383-0516  4/22/1973   156.9      61   
500       816-223-6007JinkedeKeizer@teleworm.us  1/13/1971   171.2      67   
479    TapaArsanukayev@dayrep.com1 909 458 2515  9/15/1955   220.0      65   
301  OnyekachukwuObinna@teleworm.us646-982-6609  1/24/1997   154.7      65   

      bmi  
3    31.7  
139  29.6  
500  26.8  
479  36.6  
301  25.7

In [15]:
patients.surname.value_counts()

Doe            6
Jakobsen       3
Taylor         3
Cabrera        2
Ogochukwu      2
Silva          2
Collins        2
Gersten        2
Nilsen         2
Lâm            2
Parker         2
Bùi            2
Correia        2
Lương          2
Tạ             2
Aranda         2
Cindrić        2
Kowalczyk      2
Tucker         2
Grímsdóttir    2
Souza          2
Dratchev       2
Hueber         2
Johnson        2
Batukayev      2
Liễu           2
Berg           2
Woźniak        2
Kadyrov        2
Schiavone      2
              ..
Freud          1
Geisler        1
O'Brian        1
Rokavc         1
Tar            1
Adamski        1
Solberg        1
Rap            1
Tobeolisa      1
Koldenhof      1
Karlsen        1
Gegič          1
Woodward       1
Heilmann       1
Phạm           1
Morita         1
Sleiman        1
Sørensen       1
Quintanilla    1
Filípek        1
Nordin         1
Higa           1
Jensen         1
Gormanston     1
Lê             1
Gunnarsson     1
Bois           1
Jephcott      

In [16]:
patients.address.value_counts()

123 Main Street               6
2476 Fulton Street            2
648 Old Dear Lane             2
2778 North Avenue             2
4223 Chestnut Street          1
592 Rafe Lane                 1
1624 Rosebud Avenue           1
153 Fieldcrest Road           1
1497 Hidden Meadow Drive      1
283 Simons Hollow Road        1
1051 Bassell Avenue           1
4110 Capitol Avenue           1
3391 Marcus Street            1
479 Elmwood Avenue            1
2246 Pheasant Ridge Road      1
1168 Stout Street             1
1782 New Creek Road           1
2784 Turkey Pen Road          1
2636 Benson Park Drive        1
2074 Parrish Avenue           1
377 Norman Street             1
4320 Rardin Drive             1
1886 Bicetown Road            1
208 Birch  Street             1
2310 Overlook Drive           1
1812 Poplar Street            1
4789 Devils Hill Road         1
2335 Webster Street           1
945 Maple Avenue              1
2687 Black Oak Hollow Road    1
                             ..
3964 Wal

In [17]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [18]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [10]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [11]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [7]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

In [15]:
patients_clean.shape

(503, 14)

In [16]:
treatments_clean.shape

(280, 7)

In [17]:
adverse_reactions_clean.shape

(34, 3)

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define
*Your definition here. Note: the missing `treatments` records are stored in a file named `treatments_cut.csv`, which you can see in this Jupyter Notebook's dashboard (click the **jupyter** logo in the top lefthand corner of this Notebook). Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for the function used in the solution.*

auralin and novodra have dash instead of nan vlaues.
auralin and novodra are values not variables name.

##### Code

In [8]:
# Your cleaning code here
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_clean = pd.concat([treatments_clean, treatments_cut], ignore_index = True)


In [88]:
treatments_clean

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
320      eufemio       rosario          -  37u - 40u         7.54       7.26   
321     dalmacia        madrid          -  26u - 23u         7.67       7.21   
322        daimy         tromp          -  40u - 45u         9.41       8.94   
323   hans peter      radislav  42u - 53u          -         7.60       7.20   
324       jeremy       montagu          -  52u - 52u         7.68       7.36   
325    štefanija       novosel  37u - 42u          -         7.64       7.17   
326         anja       mueller  43u - 56u          -         7.58       7.29   
327       regolo         nucci  51u - 59u          -         7.53       7.02   
328      nebechi     ekechukwu          -  37u - 39u         7.78       7.39   
329      satsita     batukayev          -  42u - 42u         7.63       7.25   
330      timothy        cotton          -  26u - 25u         7.92       7.52   
331      bjørnar        nilsen          -  36u - 33u         7.99       7.70   
332       vivien        zsinkó  30u - 38u          -         8.35       7.92   
333        genet      fesahaye  25u - 35u          -         7.73       7.36   
334        borna     šlezinger          -  42u - 41u         7.55       7.18   
335         mary         adams          -  32u - 33u         7.65       7.26   
336       blanka      jurković  31u - 43u          -         7.77       7.33   
337       joseph        tucker  48u - 56u          -         7.67       7.30   
338  c

In [19]:
treatments_clean.shape

(350, 7)

##### Test

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
*Your definition here.*
We can fix this problem to subtract hba1c_end to hba1c_start.

##### Code

In [9]:
# Your cleaning code here
#treatments_clean.hba1c_change = (treatments_clean.hba1c_start - treatments_clean.hba1c_end)
                                 
treatments_clean['hba1c_change']= treatments_clean['hba1c_start'] - treatments_clean['hba1c_end']

##### Test

In [29]:
# Your testing code here
treatments_clean.hba1c_change.head()

0    0.43
1    0.47
2    0.43
3    0.35
4    0.32
Name: hba1c_change, dtype: float64

### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
*Your definition here. Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern]
(http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

phone numbers have dash, space and dot, we phone numbers in same format and extract phone numbers and emails from the contact column.

##### Code

In [21]:
patients_clean.contact

0                   951-719-9170ZoeWellish@superrito.com
1                  PamelaSHill@cuvox.de+1 (217) 569-3204
2                       402-363-6804JaeMDebord@gustr.com
3             PhanBaLiem@jourrapide.com+1 (732) 636-8246
4                        334-515-7487TimNeudorf@cuvox.de
5               386-334-5237RafaelCardosoCosta@gustr.com
6                      775-533-5933MaryBAdams@einrot.com
7                   XiuxiuChang@einrot.com1 408 778 3236
8                816-265-9578DavidGustafsson@armyspy.com
9          SophieCabreraIbarra@teleworm.us1 718 795 9124
10                906-478-8949SandyGunnarsson@dayrep.com
11            Abdul-NurMummarIsa@rhyta.com1 931 207 0839
12             OmeokachieIbeamaka@einrot.com434-509-2614
13            AnenechiChidi@armyspy.com+1 (205) 417-8095
14                     AsiaWozniak@rhyta.com918-712-3469
15                       276-225-1955SrenFLund@gustr.com
16                LieuThiThuTam@dayrep.com1 559 765 7836
17              RoxanneAndreyev

In [10]:
# Your cleaning code here
import re
#patients_clean['phone_number'] = patients_clean.contact.str.extract('(^[1|+]?\d{0,3}?\s?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}', expand = True)
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand = True)


In [11]:
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9.-]+[a-zA-Z])', expand = True)

##### Test

In [12]:
patients_clean.drop('contact', axis = 1, inplace = True)

In [13]:
list(patients_clean)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

##### Code

In [34]:
# Your cleaning code here
# id_vars = ['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change']#we don't want to change these columns
# treatments_clean = pd.melt(treatments_clean, id_vars = id_vars, var_name = 'treatment', value_name = 'dose')

In [22]:
treatments_clean.head(1)

given_name   surname    auralin novodra  hba1c_start  hba1c_end  \
0   veronika  jindrová  41u - 48u       -         7.63        7.2   

   hba1c_change  
0          0.43

In [14]:
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str
treatments_clean = treatments_clean.drop('dose', axis=1)
# treatments_clean = treatments_clean[treatments_clean['dose']!= '-']
# treatments_clean

In [36]:
#sort treatments by patients' surname
#treatments_clean.sort_values(by = ['surname'], inplace = True)

In [37]:
#re-index drop the old index
#treatments_clean.reset_index(drop = True, inplace = True)

##### Test

In [15]:
# Your testing code here
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u  
7        31u      38u  
9        30u      36u

In [39]:
treatments_clean.dtypes

given_name       object
surname          object
hba1c_start     float64
hba1c_end       float64
hba1c_change    float64
treatment        object
dose_start       object
dose_end         object
dtype: object

#### Adverse reaction should be part of the `treatments` table

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

merging the treatments_clean table to the adverse_reaction on given name and surname.

##### Code

In [16]:
# Your cleaning code here
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on = ['given_name', 'surname'], how = 'left')

##### Test

In [24]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u  
7        31u      38u  
9        30u      36u

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

Adverse reactions table is no longer needed so ignore that part. Isolate the patient ID and names in the patients table, then convert these names to lower case to join with treatments. Then drop the given name and surname columns in the treatments table (so these being lowercase isn't an issue anymore).

##### Code

In [17]:
# Your cleaning code here
id_names = patients_clean[['patient_id', 'given_name', 'surname']]
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()
id_names.head(1)

/opt/conda/lib/python3.6/site-packages/pandas/core/generic.py:4405: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


patient_id given_name  surname
0           1        zoe  wellish

In [18]:
treatments_clean = pd.merge(treatments_clean, id_names, on = ['given_name', 'surname'])
treatments_clean.head(1)

given_name   surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika  jindrová         7.63        7.2          0.43   auralin   

  dose_start dose_end adverse_reaction  patient_id  
0        41u      48u              NaN         225

In [44]:
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis = 1)
                                  

##### Test

In [45]:
# Your testing code here
treatments_clean

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0           7.63       7.20          0.43   auralin        41u      48u   
1           7.97       7.62          0.35   auralin        33u      36u   
2           7.65       7.27          0.38   auralin        37u      42u   
3           7.89       7.55          0.34   auralin        31u      38u   
4           7.76       7.37          0.39   auralin        30u      36u   
5           7.70       7.19          0.51   auralin        29u      36u   
6           7.70       7.19          0.51   auralin        29u      36u   
7           9.54       9.14          0.40   auralin        29u      38u   
8           7.74       7.30          0.44   auralin        27u      37u   
9           7.78       7.34          0.44   auralin        55u      68u   
10          7.53       7.13          0.40   auralin        28u      37u   
11          7.61       7.29          0.32   auralin        29u      39u   
12          8.61       8.18          0.43   auralin        53u      60u   
13          9.68       9.29          0.39   auralin        31u      41u   
14          7.79       7.40          0.39   auralin        42u      51u   
15          7.81       7.48          0.33   auralin        42u      49u   
16          7.70       7.38          0.32   auralin        35u      39u   
17          7.96       7.55          0.41   auralin        47u      58u   
18          7.68       7.24          0.44   auralin        45u      48u   
19          7.92       7.47          0.45   auralin        24u      37u   
20          7.92       7.57          0.35   auralin        44u      55u   
21          7.53       7.15          0.38   auralin        37u      43u   
22          7.67       7.37          0.30   auralin        43u      47u   
23          7.86       7.51          0.35   auralin        36u      42u   
24          9.18       8.64          0.54   auralin        29u      37u   
25          7.61       7.16          0.45   auralin        57u      64u   
26          7.52       7.11          0.41   auralin        54u      67u   
27          7.71       7.34          0.37   auralin        34u      42u   
28          7.87       7.47          0.40   auralin        29u      37u   
29          7.74       7.32          0.42   auralin        61u      67u   
..           ...        ...           ...       ...        ...      ...   
319         7.98       7.60          0.38   novodra        37u      38u   
320         7.61       7.14          0.47   novodra        28u      24u   
321         7.99       7.56          0.43   novodra        24u      26u   
322         7.73       7.34          0.39   novodra        20u      21u   
323         7.71       7.30          0.41   novodra        46u      46u   
324         7.54       7.17          0.37   novodra        26u      27u   
325         7.51       7.08          0.43   novodra        34u      33u   
326         7.89       7.55          0.34   novodra        30u      35u   
327         7.71       7.30          0.41   novodra        33u      33u   
328         8.00       7.50          0.50   novodra        35u      34u   
329         7.79       7.45          0.34   novodra        30u      27u   
330         7.70       7.36          0.34   novodra        37u      41u   
331         7.84       7.37          0.47   novodra        43u      40u   
332         8.90       8.57          0.33   novodra        34u      34u   
333         7.64       7.24          0.40   novodra        40u      36u   
334         7.54       7.26          0.28   novodra        37u      40u   
335         7.67       7.21          0.46   novodra        26u      23u   
336         9.41       8.94          0.47   novodra        40u      45u   
337         7.68       7.36          0.32   novodra        52u      52u   
338         7.78       7.39          0.39   novodra        37u      39u   
339         7.63       7.25          0.38   novodra        42u      42u   
340         7.92       7.52   

In [46]:
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
Convert the zip code column's data type from a float to a string using astype, remove the '.0' using string slicing, and pad four digit zip codes with a leading 0.

##### Code

In [19]:
# Your cleaning code here
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar='0')
# Reconvert NaNs entries that were converted to '0000n' by code above
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)


##### Test

In [20]:
# Your testing code here.
patients_clean.zip_code.head()

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*Your definition here.*

##### Code

In [21]:
# Your cleaning code here
#patients_clean.query('height == 27').replace(to_value = 27, value = 72, inplace = True)
patients_clean.height = patients_clean.height.replace(27, 72)

##### Test

In [22]:
# Your testing code here
patients_clean[patients_clean['height']==27]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [23]:
patients_clean[patients_clean['surname']=='Neudorf']

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      72  26.1   

   phone_number                email  
4  334-515-7487  TimNeudorf@cuvox.de

#### Full state names sometimes, abbreviations other times

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*

##### Code

In [24]:
# Your cleaning code here, California, Illinois, Nebraska, Florida, New York
state_abbrev = {'California':'CA', 'Illinois':'IL', 'Nebraska': 'NE', 'Florida': 'FL', 'New York': 'NY'}

def abbreviate_state(patient):
    if patient['state'] in state_abbrev.keys():
        abbrev = state_abbrev[patient['state']]
        return abbrev
    else:
        return patient['state']
    
patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)        

##### Test

In [25]:
# Your testing code here
patients_clean.state.head()

0    CA
1    IL
2    NE
3    NJ
4    AL
Name: state, dtype: object

#### Dsvid Gustafsson

##### Define
*Your definition here.*

##### Code

In [26]:
# Your cleaning code here
patients_clean['given_name'] = patients_clean['given_name'].str.replace('Dsvid', 'David')

##### Test

In [27]:
# Your testing code here
patients_clean[patients_clean['given_name']=='David']

patient_id assigned_sex given_name     surname               address  \
8             9         male      David  Gustafsson    1790 Nutter Street   
280         281         male      David    Beauvais    98 Freshour Circle   
440         441         male      David   Villadsen  1434 Oak Ridge Drive   

             city state zip_code        country   birthdate  weight  height  \
8     Kansas City    MO    64105  United States    3/6/1937   163.9      66   
280   San Antonio    TX    78258  United States    5/1/1931   128.5      67   
440  St Elizabeth    MO    65559  United States  11/30/1924   137.3      68   

      bmi       phone_number                        email  
8    26.5       816-265-9578  DavidGustafsson@armyspy.com  
280  20.1       210-218-3477     DavidLBeauvais@rhyta.com  
440  20.9  +1 (573) 493-4748  DavidMVilladsen@armyspy.com

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

##### Code

In [56]:
patients_clean.dtypes

patient_id        int64
assigned_sex     object
given_name       object
surname          object
address          object
city             object
state            object
zip_code         object
country          object
birthdate        object
weight          float64
height            int64
bmi             float64
phone_number     object
email            object
dtype: object

In [57]:
treatments_clean.dtypes

hba1c_start         float64
hba1c_end           float64
hba1c_change        float64
treatment            object
dose_start           object
dose_end             object
adverse_reaction     object
patient_id            int64
dtype: object

In [28]:
# To category
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

# To datetime
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

# Strip u and to integer
treatments_clean.dose_start = treatments_clean.dose_start.str.strip('u').astype(int)
treatments_clean.dose_end = treatments_clean.dose_end.str.strip('u').astype(int)

##### Test

In [59]:
# Your testing code here
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        491 non-null object
country         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone_number    491 non-null object
email           491 non-null object
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(8)
memory usage: 53.9+ KB


In [60]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
hba1c_start         349 non-null float64
hba1c_end           349 non-null float64
hba1c_change        349 non-null float64
treatment           349 non-null object
dose_start          349 non-null int64
dose_end            349 non-null int64
adverse_reaction    35 non-null object
patient_id          349 non-null int64
dtypes: float64(3), int64(3), object(2)
memory usage: 24.5+ KB


#### Multiple phone number formats

##### Define
*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

patients_clean we need to remove space, '()', '-', and '+' so we can have all phone number in same format.

##### Code

In [29]:
# Your cleaning code here
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(width = 11, fillchar = '1')

##### Test

In [30]:
# Your testing code here
patients_clean.phone_number.head()

0    19517199170
1    12175693204
2    14023636804
3    17326368246
4    13345157487
Name: phone_number, dtype: object

#### Default John Doe data

##### Define
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*

##### Code

In [31]:
# surname = patients_clean['surname']== 'Doe'
# patients_clean.drop(index = patients_clean[surname].index, inplace = True)
doe = patients_clean.query('surname == "Doe"').index
doe

Int64Index([215, 229, 237, 244, 251, 277], dtype='int64')

In [64]:
patients_clean.drop(doe, inplace = True)

##### Test

In [65]:
# Your testing code here
patients_clean.query('surname == "Doe"')

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
*Your definition here.*

##### Code

In [66]:
# Your cleaning code here
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [67]:
# Your testing code here
patients_clean[patients_clean['surname']=='Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi phone_number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

In [68]:
patients_clean[patients_clean['surname']=='Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

    phone_number                       email  
131  13044382648    SandraCTaylor@dayrep.com  
426  13054346299  RogelioJTaylor@teleworm.us

In [69]:
patients_clean[patients_clean.surname == 'Gersten']

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    Patrick  Gersten  2778 North Avenue  Burr   

   state zip_code        country  birthdate  weight  height   bmi  \
97    NE    68324  United States 1954-05-03   138.2      71  19.3   

   phone_number                     email  
97  14028484923  PatrickGersten@rhyta.com

#### kgs instead of lbs for Zaitseva weight

##### Define
*Your definition here.*

##### Code

In [70]:
# Your cleaning code here
#patients_clean.query('weight == 48.8').replace({'weight':48.8}, 107.8, inplace = True)
weight_kg = patients_clean.weight.min()
mask = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
patients_clean.loc[mask, column_name]= weight_kg*2.20462

##### Test

In [71]:
# Your testing code here
patients_clean.query('surname == "Zaitseva"')

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate      weight  height  \
210  Wooster    OH    44691  United States 1938-11-26  107.585456      63   

      bmi phone_number                          email  
210  19.1  13302022145  CamillaZaitseva@superrito.com

In [72]:
after_adverse_reaction = treatments_clean.groupby(['treatment', 'adverse_reaction'])
after_adverse_reaction


In [73]:
treatments_clean.head(1)

hba1c_start  hba1c_end  hba1c_change treatment  dose_start  dose_end  \
0         7.63        7.2          0.43   auralin          41        48   

  adverse_reaction  patient_id  
0              NaN         225